### PR TITLE
8183521: Unable to type characters with tilde with swiss german keyboard layout

### PR DESCRIPTION
**Pull request by beldenfox**
The glass code on Windows does its own dead key processing so at certain points it must clear the dead key state that the OS is maintaining. It does this by simulating a SPACE key press but this only works reliably if the SPACE key is pressed without any modifiers.

Currently the code is picking up the current modifier state of the keyboard. If the dead key is invoked using AltGr the code is trying to clear the dead key state using AltGr+Space. This does nothing. The fix is to ignore the current keyboard state and pretend the only key being held down is Space.

I haven't figured out a way of providing automated tests for dead keys. There's no way to detect whether the correct keyboard layout is active or to force the correct layout to be active.

To test this you'll need to install the correct layout. On Windows 11 go to Settings > Time & language > Language & region. Click on the three dots to the far right of the English entry (or whatever language you're using) and select “Language options”. Scroll down until you find “Add a keyboard”. From that button scroll through all the many, many layouts until you find “Swiss German”. Once you’ve added the keyboard you should see a button on the far right of the task bar to select the layout. The dead key for this bug is where the =/+ key is on the standard US keyboard, top row to the left of Backspace.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8183521](https://bugs.openjdk.org/browse/JDK-8183521): Unable to type characters with tilde with swiss german keyboard layout (**Bug** - P4)

### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1585/head:pull/1585` \
`$ git checkout pull/1585`

Update a local copy of the PR: \
`$ git checkout pull/1585` \
`$ git pull https://git.openjdk.org/jfx.git pull/1585/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1585`

View PR using the GUI difftool: \
`$ git pr show -t 1585`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1585.diff">https://git.openjdk.org/jfx/pull/1585.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1585#issuecomment-2383675911)

### CODE REVIEW

**Comment by andy-goryachev-oracle** [comment]
@kevinrushforth could you please take a look?

File: modules/javafx.graphics/src/main/native-glass/win/ViewContainer.cpp [hunk 1]
```diff
@@ -547,8 +547,8 @@ void ViewContainer::HandleViewDeadKeyEvent(HWND hwnd, UINT msg, WPARAM wParam, L
     }
 
     // Since we handle dead keys ourselves, reset the keyboard dead key status (if any)
-    static BYTE kbState[256];
-    ::GetKeyboardState(kbState);
+    static BYTE kbState[256] = {};
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/native-glass/win/ViewContainer.cpp:550]: my c++ is a bit rusty, do we need `static` here?

Comment (beldenfox) [reply modules/javafx.graphics/src/main/native-glass/win/ViewContainer.cpp]: There's no point in this being static. I've updated the PR.

**Comment by andy-goryachev-oracle** [comment]
I've set up the Swiss German keyboard on my win11 dell laptop.
However, using + a combination results in **â** instead of **ã** per bug report

```
KeyEvent{type=KEY_PRESSED, character=<\u00>, text=^, code=DEAD_CIRCUMFLEX}
KeyEvent{type=KEY_PRESSED, character=<\u00>, text=a, code=A}
KeyEvent{type=KEY_TYPED, character=â, text=, code=UNDEFINED}
KeyEvent{type=KEY_RELEASED, character=<\u00>, text=a, code=A}
```

typing "shift, +, a" results in **à**

```
KeyEvent{type=KEY_PRESSED, character=<\u00>, text=, code=SHIFT, shift}
KeyEvent{type=KEY_PRESSED, character=<\u00>, text=`, code=DEAD_CIRCUMFLEX, shift}
KeyEvent{type=KEY_RELEASED, character=<\u00>, text=, code=SHIFT}
KeyEvent{type=KEY_PRESSED, character=<\u00>, text=a, code=A}
KeyEvent{type=KEY_TYPED, character=à, text=, code=UNDEFINED}
KeyEvent{type=KEY_RELEASED, character=<\u00>, text=a, code=A}
```

**Comment by andy-goryachev-oracle** [comment]
FYI: I've updated the monkey tester keyboard event viewer (Tools -> Keyboard Event Viewer) to improve its output, just in case.

**Comment by beldenfox** [comment]
With the Swiss German layout are three dead keys on that key (on a US layout it's the =/+ key just to the left of backspace).
- Without modifiers it's a circumflex
- With Shift it's a grave accent
- With AltGr it's a tilde
It's the AltGr case that was failing.
For testers who aren't familiar with European layouts AltGr is the Alt key to the right of the spacebar. Or so I've been told. I work on a Mac so it was easiest to program a macro key to be AltGr.

**Comment by andy-goryachev-oracle** [comment]
thanks for clarification!

on my dell laptop right alt key is a regular alt and it works as alt.  I assume it's AltGr on the actual Swiss German keyboard.

I am afraid I cannot really test this.  Could someone with AltGr keyboard help please?

**Comment by beldenfox** [comment]
If you don't have a dedicated AltGr key you can try Ctrl + Alt. It works for me on my emulated Windows 11 system with the Swiss German layout.

**Review by kevinrushforth — APPROVED** [review]
On Windows 11, I added the German Swiss keyboard layout. With that, I am able to enter AltGr DeadKey Space using the right-ALT for AltGr and the +/= key next to the backspace as the DeadKey.

I confirm the bug without this patch: I get `~^` instead of the expected `~`. It works correctly with the patch.

**Comment by kevinrushforth** [comment]
Likewise, if I type "AltGr + DeadKey + a" I get:

Without this patch: `ẫ`
With this patch: `ã`

**Comment by andy-goryachev-oracle** [comment]
Ctrl-Alt does not work on my Dell, using the embedded or external USB keyboard and Swiss German layout.

**Review by andy-goryachev-oracle — APPROVED** [review]

**Pull request closed without merging**